server/src/storage/segment.rs

```diff
@@ -104,58 +116,109 @@ impl SegmentBuilder {
+        let timestamps: &TimestampArray = data.column(0).as_any().downcast_ref().unwrap();
+        let path = format!("{}/{}.parquet", folder_path, timestamps.value(0));
+
+        write_batch_to_parquet(data, path.clone());
+
+        Self { path }
+    }
+}
+
+impl UncompressedSegment for BufferedSegment {
+    /// Retrieve the data from the Parquet file and return it in a structured record batch.
+    fn get_record_batch(&mut self) -> RecordBatch {
+        let file = File::open(&self.path).unwrap();
+        let file_reader = SerializedFileReader::new(file).unwrap();
+
+        // Specify that we want to read the first two columns (timestamps, values) from the file.
+        let file_metadata = file_reader.metadata().file_metadata();
+        let mask = ProjectionMask::leaves(file_metadata.schema_descr(), [0, 1]);
+
+        // Convert the read data into a structured record batch.
+        let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(file_reader));
+        let mut record_batch_reader = arrow_reader
+            .get_record_reader_by_columns(mask, 2048)
+            .unwrap();
 
-        let path = format!("{}/{}.parquet", folder_name, self.first_timestamp);
-        write_batch_to_parquet(batch, path);
+        record_batch_reader.next().unwrap().unwrap()
+    }
+
+    /// Return 0 since the data is not kept in memory.
+    fn get_memory_size(&self) -> usize {
+        0
     }
 }
 
-/// Write `batch` to an Apache Parquet file at the location given by `path`.
-fn write_batch_to_parquet(batch: RecordBatch, path: String) {
-    // Write the record batch to the parquet file buffer.
-    let file = File::create(path).unwrap();
-    let props = WriterProperties::builder()
-        .set_dictionary_enabled(false)
-        // TODO: Test using more efficient encoding. Plain encoding makes it easier to read the files externally.
-        .set_encoding(Encoding::PLAIN)
-        .build();
-    let mut writer = ArrowWriter::try_new(file, batch.schema(), Some(props)).unwrap();
-
-    writer.write(&batch).expect("Writing batch.");
-    writer.close().unwrap();
+/// Representing either an in-memory or buffered segment that is finished and ready for compression.
+pub struct FinishedSegment {
+    pub key: String,
+    pub uncompressed_segment: Box<dyn UncompressedSegment>,
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
     use paho_mqtt::Message;
 
+    // Tests for SegmentBuilder.
     #[test]
-    fn test_can_get_length() {
-        let (data_point, mut segment_builder) = get_empty_segment_builder();
+    fn test_get_segment_builder_memory_size() {
+        let mut segment_builder = SegmentBuilder::new();
+
+        let expected = (segment_builder.timestamps.capacity() * mem::size_of::<Timestamp>())
+            + (segment_builder.values.capacity() * mem::size_of::<Value>());
+
+        assert_eq!(SegmentBuilder::get_memory_size(), expected)
+    }
+
+    #[test]
+    fn test_get_length() {
+        let data_point = get_data_point();
```

Comment:
The purpose of `data_point` is unclear in this test.

Reply:
There was no purpose, it has been removed.